internal/collector/processor/nodebatcher/node_batcher.go

```diff
@@ -135,154 +130,83 @@ func (b *batcher) genBucketID(node *commonpb.Node, resource *resourcepb.Resource
 		b.tickers[rand.Intn(len(b.tickers))].add(bucketID)
 	}
-	return bucket.(*nodeBatcher)
+	return bucket.(*nodeBatch)
 }
 
 func (b *batcher) removeBucket(bucketID string) {
 	stats.Record(context.Background(), statNodesRemovedFromBatches.M(1))
 	b.buckets.Delete(bucketID)
 }
 
-type nodeBatcher struct {
-	// Please keep this field as the first element to ensure alignment for atomics on 32-bit systems
-	lastSent int64
-
-	timeout       time.Duration
-	sendBatchSize uint32
-	currBatch     *batch
-	node          *commonpb.Node
-	resource      *resourcepb.Resource // TODO(skaris) remove when resource is added to span
-	spanFormat    string
-	logger        *zap.Logger
-
-	dead            uint32
+type nodeBatch struct {
+	mu              sync.RWMutex
+	items           []*tracepb.Span
 	cyclesUntouched uint32
-	parent          *batcher
+	dead            uint32
+	lastSent        int64
+
+	parent   *batcher
+	format   string
+	node     *commonpb.Node
+	resource *resourcepb.Resource
 }
 
-func newNodeBucket(
+func newNodeBatch(
 	parent *batcher,
+	format string,
 	node *commonpb.Node,
 	resource *resourcepb.Resource,
-	spanFormat string,
-	sendBatchSize uint32,
-	timeout time.Duration,
-	logger *zap.Logger,
-) *nodeBatcher {
-	nb := &nodeBatcher{
-		timeout:       timeout,
-		sendBatchSize: sendBatchSize,
-		currBatch:     newBatch(initialBatchCapacity, sendBatchSize),
-		node:          node,
-		resource:      resource,
-		spanFormat:    spanFormat,
-		lastSent:      int64(0),
-		logger:        logger,
-
-		dead:            uint32(0),
-		cyclesUntouched: uint32(0),
-		parent:          parent,
+) *nodeBatch {
+	return &nodeBatch{
+		parent:   parent,
+		format:   format,
+		node:     node,
+		resource: resource,
+		items:    make([]*tracepb.Span, 0, initialBatchCapacity),
 	}
-
-	return nb
 }
 
-func (nb *nodeBatcher) add(spans []*tracepb.Span) {
-	// Reset cycles untouched to 0
-	atomic.StoreUint32(&nb.cyclesUntouched, 0)
-	// Add will always add these spans to the current batch, if ok is false it means that
-	// we need to cut
-	var b *batch
-	closed := true
-	cutBatch := false
-	for closed {
-		// atomic.LoadPointer only takes unsafe.Pointer interfaces. We do not use unsafe
-		// to skirt around the golang type system.
-		b = (*batch)(atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&nb.currBatch))))
-		cutBatch, closed = b.add(spans)
-	}
-	if atomic.LoadUint32(&nb.dead) == nodeStatusDead || cutBatch {
-		statsTags := processor.StatsTagsForBatch(
-			nb.parent.name, processor.ServiceNameForNode(nb.node), nb.spanFormat,
-		)
-		if cutBatch {
-			stats.RecordWithTags(context.Background(), statsTags, statBatchSizeTriggerSend.M(1))
-		} else {
-			stats.RecordWithTags(context.Background(), statsTags, statBatchOnDeadNode.M(1))
-		}
-		// Shoot off event so that we don't block this add on other adds
-		go nb.cutBatch(b)
-	}
-}
+func (nb *nodeBatch) add(spans []*tracepb.Span) {
+	nb.mu.Lock()
+	nb.items = append(nb.items, spans...)
```

Comment:
spans []*tracepb.Span should be immutable so why not make `items [][]*tracepb.Span` then this will become `nb.items = append(nb.items, spans[:len(spans)])` or just `nb.items = append(nb.items, spans)`

Reply:
I think we would end up having to loop through to make a single list at the end to pass along since that is what `data.TraceData` requires. Idk if you know a good/more efficient way to collapse these lists

Reply:
Indeed you need to do that but you don't do that while holding the lock. So you avoid contention.

Reply:
ah, makes sense, will do

Reply:
Are you planning to do this?

Reply:
Yes, it should be done in the most recent code